samza-yarn/src/main/java/org/apache/samza/util/hadoop/CertFSFileSystem.java

```diff
@@ -0,0 +1,461 @@
+        "-----BEGIN RSA PRIVATE KEY-----",
+        encoder.encodeToString(privateKey),
+        "-----END RSA PRIVATE KEY-----"
+    );
+    return privateKeyPem;
+  }
+
+  /**
+   * get https client based on the configuration settings
+   * @return HttpClient
+   * @throws CertFSException when there is any exception when creating httpClient for SSL
+   */
+  private HttpClient getHttpsClient() throws CertFSException {
+    String trustStoreType = getConf().get(CertFSConstants.TRUSTSTORE_TYPE, "JKS");
+    String keyStoreType = getConf().get(CertFSConstants.KEYSTORE_TYPE, "PKCS12");
+    try {
+      KeyStore trustStore = KeyStore.getInstance(trustStoreType);
+      KeyStore keyStore = KeyStore.getInstance(keyStoreType);
+
+      String trustStoreLocation = getConf().get(CertFSConstants.TRUSTSTORE_LOCATION, "");
+      log.info("trust store location is: {}", trustStoreLocation);
+      String trustStorePassword = getConf().get(CertFSConstants.TRUSTSTORE_PASSWORD, "");
+      trustStore.load(new FileInputStream(trustStoreLocation), trustStorePassword.toCharArray());
+      TrustManagerFactory tmf = TrustManagerFactory.getInstance(TrustManagerFactory.getDefaultAlgorithm());
+      tmf.init(trustStore);
+      TrustManager[] tms = tmf.getTrustManagers();
+
+      String keyStoreLocation = getConf().get(CertFSConstants.KEYSTORE_LOCATION, "");
+      log.info("key store location is: {}", keyStoreLocation);
+
+      String keyStorePassword = getConf().get(CertFSConstants.KEYSTORE_PASSWORD, "");
+      String keyPassword = getConf().get(CertFSConstants.KEY_PASSWORD, "");
+      keyStore.load(new FileInputStream(keyStoreLocation), keyStorePassword.toCharArray());
+      KeyManagerFactory kmf = KeyManagerFactory.getInstance(KeyManagerFactory.getDefaultAlgorithm());
+      kmf.init(keyStore, keyPassword.toCharArray());
+      KeyManager[] kms = kmf.getKeyManagers();
+
+      SSLContext sslContext = SSLContext.getInstance("SSL");
+      sslContext.init(kms, tms, new SecureRandom());
+
+      return HttpClients.custom().setSslcontext(sslContext).build();
+    } catch (Exception e) {
+      throw new CertFSException("Exception on getting https client", e);
+    }
+  }
+
+  /**
+   * send https POST request to URI
+   * @param httpsUri https uri the request is sent to
+   * @param payloadStr payload json in the https body
+   * @return response entity from the URI service
+   * @throws CertFSException
+   * @throws IOException
+   */
+  private String sendRequest(final URI httpsUri, final String payloadStr) throws CertFSException, IOException {
+    log.info("httpsUri: {} with payloadStr: {}", httpsUri, payloadStr);
+    HttpClient client = getHttpsClient();
+
+    HttpPost httpPost = new HttpPost(httpsUri.toString());
+    StringEntity params = new StringEntity(payloadStr);
+
+    httpPost.addHeader("content-type", "application/json; charset=UTF-8"); //hard coded here
+    httpPost.setEntity(params);
+    HttpResponse response = client.execute(httpPost);
```

Comment:
We just leave the CertFSException and IOException bubble to the open(). 
It is a nice catch to have timeout on the request. I will add it. thanks for catching it!